modules/core/import/modules/csv/src/Controller/CsvImportController.php

```diff
@@ -94,4 +136,126 @@ public function index(): array {
+  }
+
+  /**
+   * Controller for individual importers.
+   *
+   * @param string $migration_id
+   *   The migration ID.
+   *
+   * @return array
+   *   Returns a render array.
+   */
+  public function importer(string $migration_id): array {
+
+    // Load the migration definition.
+    $migration = $this->pluginManagerMigration->getDefinition($migration_id);
+
+    // Show column descriptions, if available.
+    if (!empty($migration['third_party_settings']['farm_import_csv']['columns'])) {
+
+      // Create a collapsed fieldset.
+      $build['columns'] = [
+        '#type' => 'details',
+        '#title' => $this->t('CSV Columns'),
+      ];
+
+      // Show a description of the columns with a link to download a template.
+      $items = [];
+      foreach ($migration['third_party_settings']['farm_import_csv']['columns'] as $info) {
+        if (!empty($info['name'])) {
+          $item = '<strong>' . $info['name'] . '</strong>';
+          if (!empty($info['description'])) {
+            $item .= ': ' . $info['description'];
+          }
+          $items[] = Markup::create($item);
+        }
+      }
+      $template_link = Link::createFromRoute($this->t('Download template'), 'farm.import.csv.template', ['migration_id' => $migration_id]);
+      $build['columns']['descriptions'] = [
+        '#theme' => 'item_list',
+        '#items' => $items,
+        '#suffix' => '<p>' . $template_link->toString() . '</p>',
+      ];
+    }
+
+    // Add the importer form.
+    $build['form'] = $this->formBuilder->getForm('Drupal\farm_import_csv\Form\CsvImportForm', $migration_id);
+
+    return $build;
+  }
+
+  /**
+   * Download a template for a CSV migration.
+   *
+   * @param string $migration_id
+   *   The migration ID.
+   *
+   * @return \Symfony\Component\HttpFoundation\Response
+   *   An application/csv file download response object.
+   */
+  public function template(string $migration_id) {
+    try {
+      /** @var \Drupal\migrate\Plugin\MigrationInterface $migration */
+      $migration = $this->pluginManagerMigration->getDefinition($migration_id);
+    }
+    catch (PluginNotFoundException $e) {
+      throw new ResourceNotFoundException();
+    }
+    if ($migration['migration_group'] != 'farm_import_csv') {
+      throw new ResourceNotFoundException();
+    }
+    else {
+      $filename = str_replace(':', '--', $migration_id) . '.csv';
+      $response = new Response();
+      $response->headers->set('Content-Type', 'application/csv');
+      $response->headers->set('Content-Disposition', 'attachment; filename="' . $filename . '"');
+      $column_names = [];
+      if (!empty($migration['third_party_settings']['farm_import_csv']['columns'])) {
+        foreach ($migration['third_party_settings']['farm_import_csv']['columns'] as $column) {
+          if (!empty($column['name'])) {
+            $column_names[] = $column['name'];
+          }
+        }
+      }
```

Comment:
Could be simplified:
```suggestion
      $column_names = array_filter(array_column($migration['third_party_settings']['farm_import_csv']['columns'] ?? [], 'name'));
```

Reply:
Ah great! Fixup on the way...